app/waf/src/controllers/cert.controller.ts

```diff
@@ -0,0 +1,270 @@
+      Cert,
+      await this.certRepository.findById(id, {
+        where: {
+          and: [
+            {or: [{tenantId: await this.tenantId}, {public: true}]},
+            {id: id},
+          ],
+        },
+      }),
+    );
+  }
+
+  @post(prefix + '/certs/{id}/adcs/{adcId}', {
+    responses: {
+      '204': Schema.emptyResponse('Uploading Cert resource'),
+      '404': Schema.notFound('Can not find trustedDeivceId'),
+      '422': Schema.unprocessableEntity(' Adc is not trusted'),
+    },
+  })
+  async certUpload(
+    @param(Schema.pathParameter('id', 'Cert resource ID')) id: string,
+    @param(Schema.pathParameter('adcId', 'ADC resource ID')) adcId: string,
+    @requestBody() certObj: ContentBody,
+  ): Promise<void> {
+    let cert: Cert = await this.certRepository.findById(id, {
+      where: {
+        and: [
+          {or: [{tenantId: await this.tenantId}, {public: true}]},
+          {id: id},
+        ],
+      },
+    });
+
+    let adc: Adc = await this.adcRepository.findById(adcId, undefined, {
+      tenantId: await this.tenantId,
+    });
+
+    if (!adc.management.trustedDeviceId) {
+      throw new HttpErrors.UnprocessableEntity(`Adc: ${adc.id} is not trusted`);
+    }
+
+    let certContent = certObj!.content || null;
+    let name = certObj!.name || 'f5cert';
+
+    let length = 0;
+    if (certContent !== null) {
+      length = certContent.length;
+    }
+
+    if (!certContent || length === 0) {
+      throw new HttpErrors.UnprocessableEntity('empty cert content');
+    }
+
+    try {
+      await this.asgMgr.uploadFile(
+        adc.management.trustedDeviceId!,
+        certContent,
+        length,
+        name,
+      );
+
+      let remoteName = certObj.name + cert.id;
+      let body = {
+        command: 'install',
+        name: `${remoteName}`,
+        'from-local-file': `/var/config/rest/downloads/${path.basename(
+          cert.localfile,
+        )}`,
+      };
+      await this.asgMgr.installCert(adc.management.trustedDeviceId!, body);
+      cert.installed = true;
+      cert.remotepath = `/Common/${remoteName}`;
+      await this.certRepository.updateById(id, cert);
+    } catch (error) {
+      throw new HttpErrors.UnprocessableEntity(
+        'upload cert to asg service failed',
+      );
+    }
+  }
+
+  @get(prefix + '/certs/{id}/adcs/{adcId}', {
+    responses: {
+      '200': 'Status of cert',
+      '404': Schema.notFound('Can not find relative resource'),
+      '422': Schema.unprocessableEntity(' Adc is not trusted'),
+    },
+  })
+  async certCheckStatus(
+    @param(Schema.pathParameter('id', 'Cert resource ID')) id: string,
+    @param(Schema.pathParameter('adcId', 'ADC resource ID')) adcId: string,
+  ): Promise<Response> {
+    let cert: Cert = await this.certRepository.findById(id, {
+      where: {
+        and: [
+          {or: [{tenantId: await this.tenantId}, {public: true}]},
+          {id: id},
+        ],
+      },
+    });
+    if (cert.length === 0) {
```

Comment:
I think this.certRepository.findById() should throw EntityNotFoundError, if no cert found

Reply:
accept this.

Reply:
Did you made any change?

Reply:
fixed it.